crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,359 @@
+    // the binary data to a buffer once it has been unpacked
+    let buffer = unpack_archive(archive)?;
+
+    // take the unpacked binary data and copy it into typst_update.part
+    let temp_exe = executable
+        .parent()
+        .unwrap_or(Path::new("./"))
+        .join("typst_update.part");
+    let mut binary_part = fs::File::create(&temp_exe)
+        .map_err(|err| eco_format!("failed to create typst_update.part: {}", err))?;
+    std::io::copy(&mut buffer.as_slice(), &mut binary_part).map_err(|err| {
+        fs::remove_file(&temp_exe).ok();
+        eco_format!("failed to write typst_update.part: {}", err)
+    })?;
+
+    // self replace the binary with the data from typst_update.part
+    self_replace::self_replace(&temp_exe).map_err(|err| {
+        fs::remove_file(&temp_exe).ok();
+        eco_format!("failed to self replace binary: {}", err)
+    })?;
+    // remove the temp typst_update.part artifact
+    fs::remove_file(&temp_exe)
+        .map_err(|err| eco_format!("failed to delete typst_update.part: {}", err))?;
+
+    // done, typst updated itself
+    println!("typst updated successfully: {}", release.name);
+
+    Ok(())
+}
+
+/// Specifies the url for the target release.
+fn target_release(version: Version) -> StrResult<Release> {
+    let url = format!(
+        "https://api.github.com/repos/{}/{}/releases/tags/v{}",
+        TYPST_GITHUB_ORG, TYPST_REPO, version
+    );
+
+    download_release(&url)
+}
+
+/// Specifies the url for the latest release.
+fn latest_release() -> StrResult<Release> {
+    let url = format!(
+        "https://api.github.com/repos/{}/{}/releases/latest",
+        TYPST_GITHUB_ORG, TYPST_REPO
+    );
+
+    download_release(&url)
+}
+
+/// Downloads and parses a GitHub release from the Typst repository.
+fn download_release(url: &str) -> StrResult<Release> {
+    match ureq::get(url).call() {
+        Ok(response) => {
+            Ok(response.into_json().map_err(|_| "unable to get json from response")?)
+        }
+        Err(ureq::Error::Status(404, _)) => {
+            // TODO: maybe make enum error type in typst::diag for this?
+            bail!("release not found")
+        }
+        // TODO: same as above
+        Err(_) => bail!("network failed"),
+    }
+}
+
+/// Sorts through the assets from a given `Release` and picks the right one
+/// for this target platform.
+///
+/// Returns a compressed archive that contains the Typst pre-compiled binary.
+fn download_asset_archive(release: &Release) -> StrResult<Archive> {
+    let asset_needed = asset_needed()?;
+
+    let asset = release
+        .assets
+        .iter()
+        .find(|a| a.name.starts_with(asset_needed))
+        .ok_or("could not find release for your target platform.")?;
+
+    let response = match ureq::get(&asset.browser_download_url).call() {
+        Ok(response) => response,
+        Err(ureq::Error::Status(404, _)) => {
+            // TODO: maybe make enum error type in typst::diag for this?
```

Comment:
Regarding the comment, I think that wouldn't be appropriate, as this is a CLI-only thing anyways (users of the Typst compiler as a library won't ever have this error).